lib/chef/provider/aws_ebs_volume.rb

```diff
@@ -68,4 +99,97 @@ def desired_options
       AWSResource.lookup_options(options, resource: new_resource)
     end
   end
+
+  def update_attachment(volume)
+    status = volume.status
+    #
+    # If we were told to attach the volume to a machine, do so
+    #
+    if desired_instance.is_a?(AWS::EC2::Instance)
+      case status
+      when :in_use
+        attachment = volume.attachments.first
+        # wrong instance
+        if attachment.instance != desired_instance
+          detach(volume, :instance => attachment.instance, :device => attachment.device)
+          attach(volume)
+        # wrong device only
+        elsif attachment.instance == desired_instance && attachment.device != new_resource.device
+          detach(volume, :device => attachment.device)
+          attach(volume)
+        end
+
+      when :available
+        attach(volume)
+      when nil
+        raise "EBS volume #{new_resource.name} does not currently exist!"
```

Comment:
If you find you're raising this multiple times, its probably better to just create a custom error:

```ruby
class EbsVolumeNotFound < RuntimeError do
  def initialize(new_resource) do
    super("EBS volume #{new_resource.name} (#{new_resource.volume_id}) does not exist")
  end
end
```